torch/_inductor/codegen/cpp.py

```diff
@@ -3605,48 +3718,162 @@ def can_fuse_vertical(self, node1, node2):
+        3. If failed, fallback to standard codegen.
         """
         kernel_group = self.kernel_group
+        generated_cpp_vec_kernel_count = metrics.generated_cpp_vec_kernel_count
+        cpp_kernel_proxy_list: List[CppKernelProxy] = []
+        nodes_list: List[List[SchedulerNode]] = []
+        assert isinstance(node, OuterLoopFusedSchedulerNode)
+
+        def try_outer_loop_fusion_with_local_buf(node: OuterLoopFusedSchedulerNode):
+            """
+            Codegen code with fused outer loop and local Buffer.
+            """
+            assert isinstance(node, OuterLoopFusedSchedulerNode)
+            cpp_kernel_proxy_list.clear()
+            nodes_list.clear()
+
+            def get_call_ranges(node: BaseSchedulerNode):
+                assert isinstance(node, (SchedulerNode, FusedSchedulerNode))
+                nodes: List[SchedulerNode] = node.get_nodes()  # type: ignore[assignment]
+                _, (group, reduction_group) = max(
+                    nodes, key=lambda x: int(x.is_reduction())
+                ).group
+                call_ranges = tuple(group) + tuple(reduction_group)
+                return call_ranges
+
+            if not all(
+                len(get_call_ranges(_node)) == node.outer_loop_fusion_depth + 1
+                for _node in node.get_outer_nodes()
+            ):
+                # Ref to the typical case of local buffer
+                # in https://github.com/pytorch/pytorch/blob/
+                # 1115a25c36340554442f28f9570abd42f0aface2/aten/src/ATen/native/cpu/SoftMaxKernel.cpp#L159
+                # where the buffer is with size of last dim and contiguous.
+                # Only support this typical case at first.
+                return False
 
-        if isinstance(node, OuterLoopFusedSchedulerNode):
-            cpp_kernel_proxy_list: List[CppKernelProxy] = []
-            nodes_list: List[List[SchedulerNode]] = []
+            local_buffers = []
+            for fused_scheduler_node in node.get_outer_nodes():
+                call_ranges = get_call_ranges(fused_scheduler_node)
+                keep_idx_id = len(call_ranges) - 1
+                scheduler_nodes = fused_scheduler_node.get_nodes()
+                for scheduler_node in scheduler_nodes:
+                    # all users inside same OuterLoopFusedSchedulerNode
+                    if not scheduler_node.is_reduction() and all(
+                        user.node in node.get_nodes() for user in scheduler_node.users
+                    ):
+                        local_buffer = LocalBuffer(
+                            scheduler_node,
+                            keep_idx_id,
+                            call_ranges[keep_idx_id],
+                            scheduler_node.node.layout.get_dtype(),
+                        )
+                        local_buffers.append(local_buffer)
+                        # At most 1 node with local buf for each OuterLoopFusedSchedulerNode
```

Comment:
I'm curious why you restrict this to one local buffer? It seems fairly straight-forward to map `load("buf0", ...)` onto `local_buffers["buf0"]`.

Reply:
Thanks for the comment. For the typical case (softmax) I met, there is only one local buffer. Will think about how to extend this implementation.